2_filtered_results/filter_data.py

```diff
@@ -0,0 +1,395 @@
+    ├── energy (list)          # Flattened energy values
+    └── forces (list)          # Flattened force vectors
+
+Output Structure
+----------------
+Creates filtered datasets in the current working directory with names based on input directory:
+- [input_dir_name]-z-score/     # Z-score filtered dataset (complete HuggingFace dataset)
+- [input_dir_name]-95thpercentile/  # Percentile filtered dataset (complete HuggingFace dataset)
+
+Each filtered dataset directory contains:
+├── dataset_info.json          # Updated dataset metadata
+├── state.json                 # Updated dataset state
+├── data-00000-of-00001.arrow  # Filtered data in Apache Arrow format
+├── rms_forces.png             # Force distribution visualization
+├── high_force_smiles.json     # List of removed high-force SMILES
+└── smiles.json                # List of remaining SMILES after filtering
+"""
+
+import json
+import pathlib
+
+import argparse
+import matplotlib.pyplot as plt
+import seaborn as sns
+import numpy as np
+import numpy.typing as npt
+from loguru import logger
+
+import descent.targets.energy
+import datasets
+
+
+def get_rms(array: npt.NDArray[np.float64]) -> float:
+    """Calculate root mean square of a numpy array.
+    Parameters
+    ----------
+    array : npt.NDArray[np.float64]
+        Input array for RMS calculation.
+    Returns
+    -------
+    float
+        Root mean square value of the array elements.
+    """
+    return np.sqrt(np.mean(array**2))
+
+
+def filter_nonparametrizable_molecules(
+    dataset: datasets.Dataset | datasets.DatasetDict
+) -> datasets.Dataset:
+    """Remove non-parametrizable molecules from dataset.
+
+    Filters the dataset to retain only molecules that can be parametrized
+    by the force field, removing any molecules that would cause errors
+    during force field parameter assignment.
+
+    Parameters
+    ----------
+    dataset : datasets.Dataset | datasets.DatasetDict
+        Input HuggingFace dataset containing molecular data. Must be a Dataset
+        (not DatasetDict) for compatibility with descent.targets.energy.extract_smiles.
+
+    Returns
+    -------
+    datasets.Dataset
+        Filtered dataset containing only parametrizable molecules.
+
+    Raises
+    ------
+    TypeError
+        If dataset is a DatasetDict instead of Dataset.
+
+    Notes
+    -----
+    Uses descent.targets.energy.extract_smiles to identify which molecules
+    can be successfully parametrized by the force field. This function requires
+    a datasets.Dataset object as input.
+    """
+    # Ensure we have a Dataset object, not DatasetDict
+    if isinstance(dataset, datasets.DatasetDict):
+        raise TypeError("Function requires datasets.Dataset, not DatasetDict")
+
+    # Ensure molecules are parametrizable
+    dataset_size = len(dataset)
+    unique_smiles = descent.targets.energy.extract_smiles(dataset)
+    dataset = dataset.filter(lambda d: d["smiles"] in unique_smiles)
```

Comment:
I don't see where in this code a check for parametrizability happens - all I see is unique-ifying the SMILES patterns? I would expect something to run SMARTS matching with the force field, filter out unsupported elements, etc.

https://github.com/SimonBoothroyd/descent/blob/7cd8062e2ff222047dfaccc6e45facf614abe9db/descent/targets/energy.py#L67-L76